File: day04/solution.py
lines = [ l.strip() for l in open('input.txt').readlines() ]
numbers = list(map(int, lines[0].split(',')))

boards = []
i = 2
while i < len(lines):
    l = []
    while i < len(lines) and lines[i] != '':
        l.append( list(map(int, lines[i].split())) )
        i += 1
    i += 1
    boards.append(l)

def won(b):
    return any( all(b[i][j] is None for j in range(5)) \
            or all(b[j][i] is None for j in range(5)) for i in range(5) )

def score(b, trigger):
    return sum( sum(k for k in l if k is not None) for l in b ) * trigger

nboards = len(boards)
won_boards = [ None ] * nboards
won_order = []
for n in numbers:
    for k, b in enumerate(boards):
        if won_boards[k] is not None: continue
        for l in b:
            if n in l:
                l[l.index(n)] = None # strike n
                if won(b):
                    won_boards[k] = n
                    won_order.append(k)


kfirst = won_order[0]
bfirst = boards[kfirst]
print('Part 1', score(bfirst, won_boards[kfirst]))
klast = won_order[-1]
blast = boards[klast]
print('Part 2', score(blast, won_boards[klast]))

File: day04/solution.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

#define BUFFER_LEN 1000
#define MAX_NUM 100
#define MAX_BOARDS 100

void strike(int board[5][5], int n)
{
    for(int i = 0; i < 5; i++)
        for(int j = 0; j < 5; j++)
            if(board[i][j] == n)
                board[i][j] = -1;
}

bool won(int board[5][5])
{
    for(int i = 0; i < 5; i++)
    {
        bool row = true;
        bool col = true;
        for(int j = 0; j < 5; j++)
        {
            row &= board[i][j] < 0;
            col &= board[j][i] < 0;
        }
        if (row || col) return true;
    }

    return false;
}

int score(int board[5][5], int trigger)
{
    int s = 0;
    for(int i = 0; i < 5; i++)
        for(int j = 0; j < 5; j++)
            if (board[i][j] >= 0)
                s += board[i][j];
    return s * trigger;
}

int main(void)
{
    FILE *fp = fopen("input.txt", "r");
    char buf[BUFFER_LEN];
    int numbers[MAX_NUM];
    int nnumbers = 0;

    fgets(buf, BUFFER_LEN, fp);

    char *p = strtok(buf, ",");
    while (p != NULL)
    {
        numbers[nnumbers++] = atoi(p);
        p = strtok(NULL, ",");
    }

    int nboards = 0;
    int boards[MAX_BOARDS][5][5];

    while (fgets(buf, BUFFER_LEN, fp) != NULL)
    {
        // blank line
        for (int i = 0; i < 5; i++)
        {
            fgets(buf, BUFFER_LEN, fp);
            char *p = strtok(buf, " ");
            int j = 0;
            while (p != NULL)
            {
                if (p[0] != '\n')
                    boards[nboards][i][j++] = atoi(p);
                p = strtok(NULL, " ");
            }
        }
        nboards++;
    }

    int first = -1;
    int first_trigger = -1;
    int last = -1;
    int last_trigger = -1;

    for(int i = 0; i < nnumbers; i++)
    {
        int n = numbers[i];
        for(int j = 0; j < nboards; j++)
        {
            if (!won(boards[j]))
            {
                strike(boards[j], n);
                if(won(boards[j]))
                {
                    if (first == -1)
                    {
                        first = j;
                        first_trigger = n;
                    }
                    last = j;
                    last_trigger = n;
                }
            }
        }
    }

    printf("Part 1 : %d\n", score(boards[first], first_trigger));
    printf("Part 2 : %d\n", score(boards[last], last_trigger));
}
